src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -0,0 +1,196 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Roslyn.Utilities;
+using AsyncCompletionData = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+using VSCompletionItem = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.AsyncCompletion
+{
+    internal sealed class CommitManager : IAsyncCompletionCommitManager
+    {
+        private static readonly IEnumerable<char> s_commitChars = ImmutableArray.Create(
+            ' ', '{', '}', '[', ']', '(', ')', '.', ',', ':',
+            ';', '+', '-', '*', '/', '%', '&', '|', '^', '!',
+            '~', '=', '<', '>', '?', '@', '#', '\'', '\"', '\\');
+
+        private static readonly AsyncCompletionData.CommitResult CommitResultUnhandled =
+            new AsyncCompletionData.CommitResult(isHandled: false, AsyncCompletionData.CommitBehavior.None);
+
+        public IEnumerable<char> PotentialCommitCharacters => s_commitChars;
+
+        /// <summary>
+        /// The method performs a preliminarily filtering of commit availability.
+        /// In case of a doubt, it should respond with true.
+        /// We will be able to cancel later in TryCommit based on VSCompletionItem item, e.g. based on CompletionItemRules.
+        /// </summary>
+        public bool ShouldCommitCompletion(
+            IAsyncCompletionSession session, 
+            SnapshotPoint location, 
+            char typedChar, 
+            CancellationToken cancellationToken)
+            => s_commitChars.Contains(typedChar);
```

Comment:
1. this implementation is not correct.  we have to return 'true'.  because you may return 'false' here even though the currently selected completion item *does* want to be committed.  
2. this method's signature is insufficient. it is not passing in the currently selected item.  Whether or not an item should be committed depends on the item itself (since the item defines rules that control this).

Reply:
Thank you, @cyrusnajmabadi! I see what you're saying.
We still are able to cancel commit in TryCommit. So, here we try to provide all potential commit characters for all Roslyn-related languages. If we'll not able to maintain them, we'll switch to return true in all cases.

---
In reply to: [222519833](https://github.com/dotnet/roslyn/pull/29016#discussion_r222519833) [](ancestors = 222519833)

Reply:
We can cancel and e.g. ask for further filtering. We have test scenarios covering this: EDitor asks us to TryCommit, we cancel, Editor calls UpdateItemList for the same trigger.

---
In reply to: [223928182](https://github.com/dotnet/roslyn/pull/29016#discussion_r223928182) [](ancestors = 223928182)